sql/moz-fx-data-shared-prod/subscription_platform_derived/stripe_subscriptions_history_v1/query.sql

```diff
@@ -0,0 +1,380 @@
+    JSON_VALUE(metadata, "$.previous_plan_id") AS previous_plan_id,
+  FROM
+    `moz-fx-data-shared-prod`.stripe_external.subscription_history_v1
+),
+subscriptions_history AS (
+  -- If a subscription's first history record has a previous_plan_id then synthesize
+  -- an extra history record to represent that previous plan period.
+  SELECT
+    * REPLACE (
+      0 AS subscription_history_row_number,
+      plan_change_date AS valid_to,
+      CAST(NULL AS TIMESTAMP) AS cancel_at,
+      FALSE AS cancel_at_period_end,
+      CAST(NULL AS TIMESTAMP) AS canceled_at,
+      CAST(NULL AS STRING) AS canceled_for_customer_at,
+      CAST(NULL AS TIMESTAMP) AS ended_at,
+      "active" AS status,
+      CAST(NULL AS TIMESTAMP) AS plan_change_date,
+      CAST(NULL AS STRING) AS previous_plan_id
+    )
+  FROM
+    base_subscriptions_history
+  WHERE
+    subscription_history_row_number = 1
+    AND plan_change_date IS NOT NULL
+    AND previous_plan_id IS NOT NULL
+  UNION ALL
+  SELECT
+    * REPLACE (
+      IF(
+        subscription_history_row_number = 1,
+        COALESCE(plan_change_date, valid_from),
+        valid_from
+      ) AS valid_from
+    )
+  FROM
+    base_subscriptions_history
+),
+subscriptions_history_with_lead_plan_metadata AS (
+  SELECT
+    *,
+    LEAD(plan_change_date) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+    ) AS lead_plan_change_date,
+    LEAD(previous_plan_id) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+    ) AS lead_previous_plan_id,
+  FROM
+    subscriptions_history
+),
+subscriptions_history_with_plan_ends AS (
+  SELECT
+    *,
+    -- A new `previous_plan_id` value means the previous row was the last row with that plan.
+    IF(
+      lead_previous_plan_id IS DISTINCT FROM previous_plan_id,
+      STRUCT(lead_previous_plan_id AS plan_id, lead_plan_change_date AS plan_ended_at),
+      (NULL, NULL)
+    ).*
+  FROM
+    subscriptions_history_with_lead_plan_metadata
+),
+subscriptions_history_with_previous_plan_ids AS (
+  -- Fill in previous `plan_id` values by looking forward to the next non-null `plan_id`.
+  SELECT
+    * EXCEPT (plan_id),
+    FIRST_VALUE(plan_id IGNORE NULLS) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+      ROWS BETWEEN
+        CURRENT ROW
+        AND UNBOUNDED FOLLOWING
+    ) AS plan_id,
+    LAST_VALUE(plan_ended_at IGNORE NULLS) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+      ROWS BETWEEN
+        UNBOUNDED PRECEDING
+        AND 1 PRECEDING
+    ) AS plan_started_at,
+  FROM
+    subscriptions_history_with_plan_ends
+),
+subscription_items AS (
+  SELECT
+    id AS subscription_item_id,
+    subscription_id,
+    plan_id,
+  FROM
+    `moz-fx-data-shared-prod`.stripe_external.subscription_item_v1
+  -- ZetaSQL requires QUALIFY to be used in conjunction with WHERE, GROUP BY, or HAVING.
```

Comment:
Is this still the case, I thought ZetaSQL usage got deprecated with: [PR#3755](https://github.com/mozilla/bigquery-etl/pull/3755)?

Reply:
Good call.  I think I'll do a separate PR though to remove all such ZetaSQL kludges in one fell swoop.

Reply:
#3898